envoy/matcher/matcher.h

```diff
@@ -161,10 +162,19 @@ class InputMatcher {
 
   /**
    * Whether the provided input is a match.
-   * @param absl::optional<absl::string_view> the value to match on. Will be absl::nullopt if the
+   * @param Matcher::MatchingDataType the value to match on. Will be absl::monostate() if the
    * lookup failed.
    */
   virtual bool match(const Matcher::MatchingDataType& input) PURE;
+
+  /**
+   * A set of data input types supported by InputMatcher.
+   * String is default supported data input type because almost all the derived objects support
+   * string only. Override this function to provide matcher specific supported data input types.
```

Comment:
Can you provide guidance on what non-default input types should look like? In the default case it's typeid of string. Is it expected to be the typeid for other things too?

Reply:
Seems they are hard to read for human. But may be we needn't to read the final string?

```
int: i
std::string: NSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
A: N12_GLOBAL__N_11AE
B: N2NA1BE
```

Reply:
Re @htuch: yes, the basically it is typeid for type of [data input result](https://github.com/envoyproxy/envoy/blob/fa5f548d05da6495c175b5c124292fa0bae215de/envoy/matcher/matcher.h#L206) and the input to [input matcher](https://github.com/envoyproxy/envoy/blob/fa5f548d05da6495c175b5c124292fa0bae215de/envoy/matcher/matcher.h#L164) (the type of both is same). For example, typeid(int), typeid(float), typeid(CustomObject) like
typeid(StreamActivation). In my next PR for CEL, [StreamActivation](https://github.com/tyxia/envoy/blob/3f6ed4811e9fc6d92198b72bedb7bfd9d6215623/source/extensions/filters/common/expr/evaluator.h#L27) is the CEL context returned from `dataInput` and used in `inputMatcher` to invoke [Evaluate](https://github.com/tyxia/cel-cpp/blob/ea339f1f73818afe5699fbef8796a4c0c617b5f7/eval/public/cel_expression.h#L49)

Re @wbpcode: Yea, the return of typeid().name is not human readable. I don't think we will read this carefully. 
As long as it is consistent way of retrieving type between dataInput and inputMatcher and it is standard output (i.e., can be looked up if people want), it should be fine.

Reply:
Technically i think it can be anything. It does not need to be typeid(). typeid() depends on RTTI being enabled, which I think at this point is a requirement for building Envoy.

I would just define a few human readable strings for core supported types, in case anyone needs to debug this in the future.

Reply:
Yea, for core types like std::string, I can define readable name. Just to confirm for custom object like CEL context, we are still using typeid? If not, then every time a new custom object is added, the name definition need to be added as well, which might not be ideal.

Reply:
To improve the naming readability. ideally, we can use util::Demangle (which seems not be available in OSS Envoy) or pretty print lib like boost/type_index.  In general, I think we should avoid hardcode if possible. But for those a few core types, it should be fine to have literal name for them
Will update.

Reply:
@tyxia I think at least update the code comments to convey what is really going on right now.

Reply:
I have updated the comment for core types. For future custom type like cel matcher, I will try to see if I can avoid hardcoded name. I am thinking of introducing demangle util via abi::__cxa_demangle(which is available in clang).
Thus I will update here again if needed when I introduce the CEL matcher